### PR TITLE
feat: controller health checks

### PR DESCRIPTION
**Pull request by limhawjia**
This PR removes the existing `healthz` endpoint and adds more specific `readyz` and `livez` endpoints, see https://kubernetes.io/docs/reference/using-api/health-checks/#api-endpoints-for-health.

In addition, it adds a readyz check for each controller run by the `ControllerManager` and `FTCManager`. The root `readyz` check aggregates the results from each individual controller.

Examples:

```console
% curl "localhost:11257/readyz?verbose"
[+]configmaps-automigration ok
[+]configmaps-federate ok
[+]configmaps-scheduler ok
[+]cronjobs.batch-automigration ok
[+]cronjobs.batch-federate ok
[+]cronjobs.batch-scheduler ok
[+]daemonsets.apps-automigration ok
[+]daemonsets.apps-federate ok
[+]daemonsets.apps-scheduler ok
[+]deployments.apps-automigration ok
[+]deployments.apps-federate ok
[+]deployments.apps-scheduler ok
[-]federatedcluster failed: reason withheld
[+]follower ok
[+]ingresses.networking.k8s.io-automigration ok
[+]ingresses.networking.k8s.io-federate ok
[+]ingresses.networking.k8s.io-scheduler ok
[+]jobs.batch-automigration ok
[+]jobs.batch-federate ok
[+]jobs.batch-scheduler ok
[+]limitranges-automigration ok
[+]limitranges-federate ok
[+]limitranges-scheduler ok
[+]namespaces-automigration ok
[+]namespaces-federate ok
[+]namespaces-scheduler ok
[+]persistentvolumeclaims-automigration ok
[+]persistentvolumeclaims-federate ok
[+]persistentvolumeclaims-scheduler ok
[+]persistentvolumes-automigration ok
[+]persistentvolumes-federate ok
[+]persistentvolumes-scheduler ok
[+]resourcequotas-automigration ok
[+]resourcequotas-federate ok
[+]resourcequotas-scheduler ok
[+]secrets-automigration ok
[+]secrets-federate ok
[+]secrets-scheduler ok
[+]serviceaccounts-automigration ok
[+]serviceaccounts-federate ok
[+]serviceaccounts-scheduler ok
[+]services-automigration ok
[+]services-federate ok
[+]services-scheduler ok
[+]statefulsets.apps-automigration ok
[+]statefulsets.apps-federate ok
[+]statefulsets.apps-scheduler ok
[+]storageclasses-automigration ok
[+]storageclasses-federate ok
[+]storageclasses-scheduler ok
[+]typeconfig ok
healthz check failed
```

```console
% curl -i "localhost:11257/readyz/configmaps-automigration"
HTTP/1.1 200 OK
Date: Thu, 04 May 2023 05:42:31 GMT
Content-Length: 2
Content-Type: text/plain; charset=utf-8
```

```console
% curl -i "localhost:11257/readyz/federatedcluster" 
HTTP/1.1 500 Internal Server Error
Content-Type: text/plain; charset=utf-8
X-Content-Type-Options: nosniff
Date: Thu, 04 May 2023 05:43:52 GMT
Content-Length: 61

internal server error: controller federatedcluster not ready
```

Livez checks consist of a simple ping check and a leader election check.

Examples

```console
% curl "localhost:11257/livez?verbose" 
[+]leaderElection ok
[+]ping ok
healthz check passed
```

### CODE REVIEW

File: cmd/controller-manager/app/controllermanager.go [hunk 1]
```diff
@@ -22,22 +22,26 @@ import (
...
+	fedleaderelection "github.com/kubewharf/kubeadmiral/pkg/controllermanager/leaderelection"
 	controllercontext "github.com/kubewharf/kubeadmiral/pkg/controllers/context"
 )
 
 const (
-	FederatedClusterControllerName = "federated-cluster-controller"
-	TypeConfigControllerName       = "type-config-controller"
-	MonitorControllerName          = "monitor-controller"
-	FollowerControllerName         = "follower-controller"
+	FederatedClusterControllerName = "federatedcluster"
```
Comment (gary-lgy) [inline cmd/controller-manager/app/controllermanager.go:38]: What about just `cluster`?

File: cmd/controller-manager/app/controllermanager.go [hunk 2]
```diff
@@ -46,22 +50,35 @@ var knownControllers = map[string]startControllerFunc{
...
+				klog.Errorf("Failed to start pprof server: %v", err)
+			}
+		}()
+	}
+
+	handler := healthcheck.NewMutableHealthCheckHandler()
+	handler.AddLivezChecker("ping", healthz.Ping)
+
+	var healthzAdaptor *leaderelection.HealthzAdaptor
+	if opts.EnableLeaderElect {
```
Comment (gary-lgy) [inline cmd/controller-manager/app/controllermanager.go:72]: Can this block be combined with the block below that also checks for `opts.EnableLeaderElect`?

File: cmd/controller-manager/app/controllermanager.go [hunk 3]
```diff
@@ -46,22 +50,35 @@ var knownControllers = map[string]startControllerFunc{
...
+		healthzAdaptor = &leaderelection.HealthzAdaptor{}
```
Comment (gary-lgy) [inline cmd/controller-manager/app/controllermanager.go:73]: Initializing `HealthzAdaptor{}` like this implies setting `maxTolerableExpiredLease` to 0 when checking. Is this intended?

File: pkg/controllermanager/leaderelection/leaderelection.go [hunk 4]
```diff
@@ -76,6 +77,7 @@ func NewFederationLeaderElector(
 		LeaseDuration: 15 * time.Second,
 		RenewDeadline: 10 * time.Second,
 		RetryPeriod:   5 * time.Second,
+		WatchDog:      healthzAdaptor,
```
Comment (gary-lgy) [inline pkg/controllermanager/leaderelection/leaderelection.go:80]: Do we need to `healthzAdaptor.SetLeaderElection(le)`?

File: cmd/controller-manager/app/controllermanager.go (hunk 2 above)
Comment (limhawjia) [reply cmd/controller-manager/app/controllermanager.go:77]: if we combine it with the block below, we may start the healthcheck server before the leaderelection check is installed.  will that be undesirable?

File: cmd/controller-manager/app/controllermanager.go [hunk 5]
```diff
@@ -46,22 +51,39 @@ var knownControllers = map[string]startControllerFunc{
...
+				Addr:              "0.0.0.0:6060",
+				ReadHeaderTimeout: time.Second * 3,
+			}
+			if err := server.ListenAndServe(); err != nil {
+				klog.Errorf("Failed to start pprof server: %v", err)
+			}
+		}()
+	}
+
+	handler := healthcheck.NewMutableHealthCheckHandler()
```
Comment (gary-lgy) [inline cmd/controller-manager/app/controllermanager.go:73]:
```suggestion
	healthCheckHandler := healthcheck.NewMutableHealthCheckHandler()
```

File: pkg/controllermanager/leaderelection/leaderelection.go (hunk 4 above)
Comment (limhawjia) [reply pkg/controllermanager/leaderelection/leaderelection.go:80]: yes, fixed

File: cmd/controller-manager/app/controllermanager.go (hunk 3 above)
Comment (limhawjia) [reply cmd/controller-manager/app/controllermanager.go]: no, fixed it

File: pkg/controllermanager/leaderelection/leaderelection.go (hunk 4 above)
Comment (gary-lgy) [reply pkg/controllermanager/leaderelection/leaderelection.go:80]: Actually, why don't we use `le.Check(maxTolerableExpiredDuration)` directly?

File: cmd/controller-manager/app/controllermanager.go (hunk 2 above)
Comment (gary-lgy) [reply cmd/controller-manager/app/controllermanager.go:77]: Why don't we run healthcheck as the last step?

File: cmd/controller-manager/app/controllermanager.go [hunk 6]
```diff
@@ -124,22 +140,33 @@ func startControllers(
...
 		}
 
-		err := initFn(ctx, controllerCtx)
+		controller, err := initFn(ctx, controllerCtx)
 		if err != nil {
 			return fmt.Errorf("error starting %q: %w", controllerName, err)
 		}
 		klog.Infof("Started %q", controllerName)
+
+		healthCheckHandler.AddReadyzChecker(controllerName, controllermanager.HealthzCheckerAdaptor(controllerName, controller))
```
Comment (gary-lgy) [inline cmd/controller-manager/app/controllermanager.go:149]: Do we need to capture the controller name in the error message? Would a function literal that calls `controller.IsControllerReady` work?

File: cmd/controller-manager/app/controllermanager.go (hunk 2 above)
Comment (limhawjia) [reply cmd/controller-manager/app/controllermanager.go:77]: `elector.Run` is run as the last step as it blocks until either `ctx` is done or the lease is lost, allowing `main` to return once the leader lease is lost.

we can still run healthcheck as the last step if you wish, but we would have to configure the `OnStoppedLeading` callback to cancel `ctx` in order to be notified when we lose the leader lease.

File: pkg/controllermanager/leaderelection/leaderelection.go (hunk 4 above)
Comment (limhawjia) [reply pkg/controllermanager/leaderelection/leaderelection.go:80]: that's what the adaptor is doing, it just wraps it in a way that is compatible with `healthz.Handler`

File: cmd/controller-manager/app/controllermanager.go (hunk 6 above)
Comment (limhawjia) [reply cmd/controller-manager/app/controllermanager.go:149]: using a function literal is fine as well

File: pkg/controllers/scheduler/scheduler.go [hunk 7]
```diff
@@ -95,6 +95,14 @@ type Scheduler struct {
 	logger  klog.Logger
 }
 
+func (s *Scheduler) IsControllerReady() bool {
+	return s.federatedObjectSynced() &&
+		s.clusterPropagationPolicySynced() &&
```
Comment (gary-lgy) [inline pkg/controllers/scheduler/scheduler.go:100]: This looks wrong... Can we extract the `cachesSynced` slice in `Run` into a `HasSynced` function and call that instead?

File: cmd/controller-manager/app/controllermanager.go (hunk 2 above)
Comment (gary-lgy) [reply cmd/controller-manager/app/controllermanager.go:77]: In that case, can we 1. create leader elector; 2. register livez; 3. run elector?

File: cmd/controller-manager/app/ftcmanager.go [hunk 8]
```diff
@@ -18,166 +18,207 @@ package app
...
 
+	needRetry := false
 	for controllerName, startFunc := range m.registeredSubControllers {
-		if !isControllerEnabled(controllerName, controllersDisabledByDefault, m.enabledControllers) {
-			keyedLogger.WithValues("controller", controllerName).Info("Skip starting disabled subcontroller")
-			continue
+		logger := logger.WithValues("controller", controllerName)
+
+		if startedSubControllers.Has(controllerName) {
+			logger.V(3).Info("Subcontroller already started")
```
Comment (gary-lgy) [inline cmd/controller-manager/app/ftcmanager.go:181]: continue?

File: cmd/controller-manager/app/ftcmanager.go [hunk 9]
```diff
@@ -18,166 +18,207 @@ package app
...
+		}
+
+		isEnabledFunc := m.isSubControllerEnabledFuncs[controllerName]
+		if isEnabledFunc != nil && !isEnabledFunc(typeConfig) {
+			logger.V(3).Info("Skip starting subcontroller, is disabled")
```
Comment (gary-lgy) [inline cmd/controller-manager/app/ftcmanager.go:186]: continue?

File: cmd/controller-manager/app/ftcmanager.go [hunk 10]
```diff
@@ -18,166 +18,207 @@ package app
...
-		// TODO[ftcmanager]: handle controllers that do not need to be started for certain FTCs
-		if err := startFunc(subControllerCtx, m.controllerCtx, typeConfig); err != nil {
-			keyedLogger.Error(err, "Failed to start subcontrolelr")
+		controller, err := startFunc(subControllerCtx, m.controllerCtx, typeConfig)
+		if err != nil {
+			logger.Error(err, "Failed to start subcontroller")
+			needRetry = true
 		} else {
-			keyedLogger.WithValues("controller", controllerName).Info("Started subcontroller")
+			logger.WithValues("controller", controllerName).Info("Started subcontroller")
```
Comment (gary-lgy) [inline cmd/controller-manager/app/ftcmanager.go:194]: `"controller"` should have already been added to logger

File: cmd/controller-manager/app/ftcmanager.go [hunk 11]
```diff
@@ -190,5 +231,16 @@ func (m *FederatedTypeConfigManager) processFTCDeletion(ftcName string) {
...
+		m.healthCheckHandler.RemoveReadyzChecker(resolveSubcontrollerName(controller, ftcName))
+	}
+
 	delete(m.subControllerCancelFuncs, ftcName)
+	delete(m.subControllerContexts, ftcName)
+	delete(m.startedSubControllers, ftcName)
+}
+
+func resolveSubcontrollerName(baseName, ftcName string) string {
+	return fmt.Sprintf("%s-%s", ftcName, baseName)
```
Comment (gary-lgy) [inline cmd/controller-manager/app/ftcmanager.go:245]: What about `ftcName[baseName]`? It should be visually more distinguishable

File: pkg/controllermanager/interfaces.go
Comment (gary-lgy) [inline pkg/controllermanager/interfaces.go:1]: can we name this file `types.go`

File: pkg/controllermanager/interfaces.go [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+	StartFunc     StartFTCSubControllerFunc
+	IsEnabledFunc IsFTCSubControllerEnabledFunc
+}
+
+// StartFTCSubControllerFunc is responsible for constructing and starting a FTC subcontroller. A FTC subcontroller is started/stopped
+// dynamically for every FTC. StartFTCSubControllerFunc should be asynchronous and an error is only returned if we fail to start the
+// controller.
+//
+//nolint:lll
+type StartFTCSubControllerFunc func(ctx context.Context, controllerCtx *controllercontext.Context, typeConfig *fedcorev1a1.FederatedTypeConfig) (Controller, error)
```
Comment (gary-lgy) [inline pkg/controllermanager/interfaces.go:48]: any reason this must be kept on one line?

File: pkg/controllers/automigration/controller.go [hunk 13]
```diff
@@ -80,6 +80,19 @@ type Controller struct {
 	logger  klog.Logger
 }
 
+// IsControllerReady implements controllermanager.Controller
+func (c *Controller) IsControllerReady() bool {
```
Comment (gary-lgy) [inline pkg/controllers/automigration/controller.go:84]: Can we reconcile this with `Run`, there's duplicated logic

File: pkg/controllers/federate/controller.go [hunk 14]
```diff
@@ -86,6 +86,10 @@ type FederateController struct {
 	logger  klog.Logger
 }
 
+func (c *FederateController) IsControllerReady() bool {
```
Comment (gary-lgy) [inline pkg/controllers/federate/controller.go:89]: Let's create a `HasSynced` method and call it in both `IsControllerReady` and `Run` so the two are always in sync.  This should also apply to all the other controllers.

**Comment by gary-lgy** [comment]
In the PR description the following example was given:

> % curl "localhost:11257/readyz?verbose"
> [+]configmaps-automigration ok
> [+]configmaps-federate ok
> [+]configmaps-scheduler ok
> [+]cronjobs.batch-automigration ok
> [+]cronjobs.batch-federate ok
> [+]cronjobs.batch-scheduler ok
> [+]daemonsets.apps-automigration ok
> [+]daemonsets.apps-federate ok
> [+]daemonsets.apps-scheduler ok
> [+]deployments.apps-automigration ok
> [+]deployments.apps-federate ok
> [+]deployments.apps-scheduler ok
> **[-]federatedcluster failed: reason withheld**
> [+]follower ok
> [+]ingresses.networking.k8s.io-automigration ok
> [+]ingresses.networking.k8s.io-federate ok
> [+]ingresses.networking.k8s.io-scheduler ok
> [+]jobs.batch-automigration ok
> [+]jobs.batch-federate ok
> [+]jobs.batch-scheduler ok
> [+]limitranges-automigration ok
> [+]limitranges-federate ok
> [+]limitranges-scheduler ok
> [+]namespaces-automigration ok
> [+]namespaces-federate ok
> [+]namespaces-scheduler ok
> [+]persistentvolumeclaims-automigration ok
> [+]persistentvolumeclaims-federate ok
> [+]persistentvolumeclaims-scheduler ok
> [+]persistentvolumes-automigration ok
> [+]persistentvolumes-federate ok
> [+]persistentvolumes-scheduler ok
> [+]resourcequotas-automigration ok
> [+]resourcequotas-federate ok
> [+]resourcequotas-scheduler ok
> [+]secrets-automigration ok
> [+]secrets-federate ok
> [+]secrets-scheduler ok
> [+]serviceaccounts-automigration ok
> [+]serviceaccounts-federate ok
> [+]serviceaccounts-scheduler ok
> [+]services-automigration ok
> [+]services-federate ok
> [+]services-scheduler ok
> [+]statefulsets.apps-automigration ok
> [+]statefulsets.apps-federate ok
> [+]statefulsets.apps-scheduler ok
> [+]storageclasses-automigration ok
> [+]storageclasses-federate ok
> [+]storageclasses-scheduler ok
> [+]typeconfig ok
> healthz check failed

Note the bolded line. Is "reason withheld" intentional?

Also, 

> % curl "localhost:11257/livez?verbose" 
> [+]leaderElection ok
> [+]ping ok
> healthz check passed

It states "healthz check passed" although it's a livez endpoint.

**Review by gary-lgy — APPROVED** [review]

**Pull request merged**